src/Lumper.UI/Services/BspService.cs

```diff
@@ -290,6 +294,30 @@ public async Task<bool> Save(IStorageFile? outFile)
                     makeBackup: MakeBackup),
                 RxApp.TaskpoolScheduler);
 
+            if (outName != FileName)
+            {
+                // update the view model to add the new keys to the UI
+                // need to do this here as otherwise the new keys will also get deleted
+                PakfileLumpViewModel.MarkAsModified();
+                PakfileLumpViewModel.UpdateEntries(true);
+                PakfileLumpViewModel.MarkAsModified();
+                PakfileLumpViewModel.UpdateModel();
+
+                // clean up old outdated keys
+                foreach (KeyValuePair<string, string> entry in modified)
+                {
+                    foreach (PakfileEntryViewModel pk in PakfileLumpViewModel?.Entries?.Items)
+                        if (pk.Key == entry.Key)
+                            PakfileLumpViewModel.DeleteEntry(pk);
+                }
+                // update model one more time
+                PakfileLumpViewModel.MarkAsModified();
+                PakfileLumpViewModel.UpdateEntries(true);
+                PakfileLumpViewModel.MarkAsModified();
+                PakfileLumpViewModel.UpdateModel();
+            }
+
+
```

Comment:
Shouldn't need to do this, just update keys on the viewmodel. The library code should just be updating keys on PakfileLump, which will cause the zip entries to get moved when the BSP is saved. Then you should be able to just loop through every pakfile entry viewmodel and update the key if changed. Should be all you need. Please ask on here or Discord if this stuff isn't make sense, I appreciate that the UI code can be hard to get your head around